ml/xgboost/src/main/java/ai/djl/ml/xgboost/XgbNDArray.java

```diff
@@ -0,0 +1,160 @@
+
+    XgbNDArray(XgbNDManager manager, long handle, Shape shape, SparseFormat format) {
+        this.handle = new AtomicLong(handle);
+        this.uid = String.valueOf(handle);
+        this.manager = manager;
+        this.manager.attachInternal(uid, this);
+        this.shape = shape;
+        this.format = format;
+    }
+
+    XgbNDArray(XgbNDManager manager, ByteBuffer data, Shape shape) {
+        this.manager = manager;
+        this.uid = UUID.randomUUID().toString();
+        this.manager.attachInternal(uid, this);
+        this.shape = shape;
+        this.data = data;
+        this.format = SparseFormat.DENSE;
+    }
+
+    /** {@inheritDoc} */
+    public long getHandle() {
+        return handle.get();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public XgbNDManager getManager() {
+        return manager;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String getUid() {
+        return uid;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public DataType getDataType() {
+        return DataType.FLOAT32;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Device getDevice() {
+        return Device.cpu();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Shape getShape() {
+        return shape;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SparseFormat getSparseFormat() {
+        return format;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public ByteBuffer toByteBuffer() {
+        if (data == null) {
+            throw new UnsupportedOperationException("Cannot obtain value from DMatrix");
+        }
+        return data;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void attach(NDManager manager) {
+        detach();
+        this.manager = (XgbNDManager) manager;
+        manager.attachInternal(getUid(), this);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void tempAttach(NDManager manager) {
+        detach();
+        NDManager original = this.manager;
+        this.manager = (XgbNDManager) manager;
+        manager.tempAttachInternal(original, getUid(), this);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void detach() {
+        manager.detachInternal(getUid());
+        manager = XgbNDManager.getSystemManager();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String toString() {
+        if (isClosed) {
+            return "This array is already closed";
+        }
+        return "ND: "
+                + getShape()
+                + ' '
+                + getDevice()
+                + ' '
+                + getDataType()
+                + '\n'
+                + Arrays.toString(toArray());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void close() {
+        if (handle != null && handle.get() != 0L) {
+            long pointer = handle.getAndSet(0L);
+            JniUtils.deleteDMatrix(pointer);
+        }
+        if (data != null) {
```

Comment:
You don't need to check null

Reply:
Then we essentially not collecting garbage here?